blockchain/genesis.go

```diff
@@ -307,16 +308,25 @@ func (g *Genesis) ToBlock(baseStateRoot common.Hash, db database.DBManager) *typ
 		db = database.NewMemoryDBManager()
 	}
 	stateDB, _ := state.New(baseStateRoot, state.NewDatabase(db), nil, nil)
+	rules := params.Rules{}
+	if g.Config != nil {
+		rules = g.Config.Rules(new(big.Int).SetUint64(g.Number))
+	}
 	for addr, account := range g.Alloc {
-		if len(account.Code) != 0 {
-			originalCode := stateDB.GetCode(addr)
+		originalCode := stateDB.GetCode(addr)
+		if _, ok := types.ParseDelegation(account.Code); ok && rules.IsPrague {
+			stateDB.SetCodeToEOA(addr, account.Code, rules)
+		} else if len(account.Code) == 0 && len(account.Storage) != 0 && rules.IsPrague {
+			stateDB.CreateEOA(addr, false, accountkey.NewAccountKeyLegacy())
+		} else if len(account.Code) != 0 {
+			stateDB.CreateSmartContractAccount(addr, params.CodeFormatEVM, rules)
```

Comment:
'non-empty storage then CreateEOA' path was counterintuitive. Please add some explanations.

```suggestion
		if _, ok := types.ParseDelegation(account.Code); ok && rules.IsPrague {
			// EOA with code. Usually SetCodeTx creates it. Unit tests may create it here in the genesis.
			stateDB.SetCodeToEOA(addr, account.Code, rules)
		} else if len(account.Code) == 0 && len(account.Storage) != 0 && rules.IsPrague {
			// Represents an EOA that had code then nullified with another SetCodeTx. Unit tests may create it here in the genesis.
			stateDB.CreateEOA(addr, false, accountkey.NewAccountKeyLegacy())
		} else if len(account.Code) != 0 {
			// Regular genesis smart contract account.
			stateDB.CreateSmartContractAccount(addr, params.CodeFormatEVM, rules)
```